### PR TITLE
Migrate bare-metal programs to newlib

### PR DESCRIPTION
**Pull request by a0u**
This begins migrating bare-metal programs away from the monolithic `syscalls.c` environment copied from riscv-tests towards a more structured SDK based on newlib.  The key component enabling newlib is [libgloss-htif](https://github.com/ucb-bar/libgloss-htif), which replaces the standard libgloss with a syscall implementation that supports HTIF.  Refer to the [README](https://github.com/ucb-bar/libgloss-htif/blob/master/README.md) for more details on the usage of this library.

Some advantages:
* Programs that do not depend on stdio functions become much smaller, as unused code is not statically linked in anymore.
* If necessary, a more complete/compliant libc is at one's disposal.

There are two major changes to the default toolchain configuration:
* Libraries are now built with the `medany` code model: This is needed for newlib and libgcc to be linked at addresses above 0x80000000.
* multilib is enabled: This feature is not strictly required but has been requested (@jerryz123).  The downside is that build times are significantly longer.

Currently, only programs in the miscellaneous `tests/` directory have been converted.  Since this is not so urgent, the software in other Chipyard submodules are left as future work, once we discuss how to best unify the bare-metal software stack.

A final note on `encoding.h`: Since riscv-pk already installs a copy of this header to the sysroot include path, I propose that Chipyard uses `#include <riscv-pk/encoding.h>` going forward.

### CODE REVIEW

**Comment by a0u** [comment]
@jerryz123: As I suspected, upstream buildroot still breaks with a multilib RISC-V toolchain.  It fails to properly handle the case that `ld-linux-*.so` are symlinks when populating its own sysroot.  In the past, I've used this patch to fix this limitation: https://github.com/sifive/buildroot/commit/feb1a7d40064f9fe4f2d8a094788207a53d9a043

But I'm considering not enabling multilib unless you have a significant use case.

**Comment by jerryz123** [comment]
> But I'm considering not enabling multilib unless you have a significant use case.

That's fine then. I'll work around this when I need to.

**Comment by a0u** [comment]
@zhemao: @NathanTP: CI has succeeded; this is ready for review.

The default toolchain is back to being non-multilib.

File: tests/Makefile [hunk 1]
```diff
@@ -1,12 +1,18 @@
 GCC=riscv64-unknown-elf-gcc
 OBJDUMP=riscv64-unknown-elf-objdump
-CFLAGS=-mcmodel=medany -std=gnu99 -O2 -fno-common -fno-builtin-printf -Wall
-LDFLAGS=-static -nostdlib -nostartfiles -lgcc
```
Comment (NathanTP) [inline tests/Makefile]: I'm going to trust you on this. Out of curiosity: does leaving out these options cause GCC to find newlib automatically? I don't see newlib mentioned explicitly anywhere in these changes.

File: tests/Makefile [hunk 2]
```diff
@@ -1,12 +1,18 @@
...
+CFLAGS= -std=gnu99 -O2 -fno-common -fno-builtin-printf -Wall
+LDFLAGS= -static
+
+include libgloss.mk
```
[the remark this answers is not in the document]
Comment (a0u) [reply tests/Makefile]: This makes it easier to test libgloss modifications locally without needing to install over the existing one.

File: tests/libgloss.mk [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# Handle libgloss-htif dependency
+ifndef libgloss
+
+ifndef GCC
+$(error GCC is not defined)
+endif
+
+libgloss_specs := htif_nano.specs
```
[the remark this answers is not in the document]
Comment (a0u) [reply tests/libgloss.mk]: gcc uses the library search path to look for spec files.  Is it possible that your environment pointed to a different toolchain without libgloss installed?

File: tests/Makefile (hunk 1 above)
Comment (a0u) [reply tests/Makefile]: The gcc spec files subsume these options.  By default, the RISC-V ELF toolchain uses newlib if `-nostdlib` is omitted.

File: tests/libgloss.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (a0u) [reply tests/libgloss.mk]: Wait, is `make` actually invoking `cc` instead of `riscv64-unknown-elf-gcc`?

[the remark this answers is not in the document]
Comment (a0u) [reply tests/libgloss.mk]: Okay, the proper command should be `make nic-loopback.riscv`.  The `nic-loopback` target causes `make` to fall back to implicit built-in rules, which I normally disable with the line:
```make
.SUFFIXES:
```

**Pull request merged**